tests/model/model_flattening.cpp

```diff
@@ -0,0 +1,623 @@
+{
+    const std::string unitsDefiniitonModel =
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+        "<model xmlns=\"http://www.cellml.org/cellml/2.0#\" xmlns:cellml=\"http://www.cellml.org/cellml/2.0#\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" name=\"units\">\n"
+        "  <units name=\"ms\">\n"
+        "    <unit units=\"second\" prefix=\"milli\"/>\n"
+        "  </units>\n"
+        "</model>\n";
+
+    const std::string modelImportingUnits =
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+        "<model name=\"importer_of_units\" xmlns=\"http://www.cellml.org/cellml/2.0#\">\n"
+        "  <import xlink:href=\"unitsdefinitions.xml\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n"
+        "    <units units_ref=\"ms\" name=\"my_ms\"/>\n"
+        "  </import>\n"
+        "</model>\n";
+
+    const std::string e =
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+        "<model xmlns=\"http://www.cellml.org/cellml/2.0#\" name=\"importer_of_units\">\n"
+        "  <units name=\"my_ms\">\n"
+        "    <unit prefix=\"milli\" units=\"second\"/>\n"
+        "  </units>\n"
+        "</model>\n";
+
+    auto parser = libcellml::Parser::create();
+
+    auto modelWithUnitsImports = parser->parseModel(modelImportingUnits);
+    auto modelUnitsDefinitions = parser->parseModel(unitsDefiniitonModel);
+
+    EXPECT_TRUE(modelWithUnitsImports->hasUnresolvedImports());
+
+    auto units = modelWithUnitsImports->units(0);
+    auto importSource = units->importSource();
+    importSource->setModel(modelUnitsDefinitions);
+
+    EXPECT_FALSE(modelWithUnitsImports->hasUnresolvedImports());
+
+    modelWithUnitsImports->flatten();
+
+    auto printer = libcellml::Printer::create();
+
+    auto a = printer->printModel(modelWithUnitsImports);
+    EXPECT_EQ(e, a);
+
+    // Check that we haven't changed the imported model.
+    auto aModelUnitsDefinitions = printer->printModel(modelUnitsDefinitions);
+    auto eModelUnitsDefinitions = printer->printModel(parser->parseModel(unitsDefiniitonModel));
+    EXPECT_EQ(eModelUnitsDefinitions, aModelUnitsDefinitions);
+}
+
+TEST(ModelFlattening, importedComponent)
+{
+    const std::string sineDefinitionModel =
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+        "<model xmlns=\"http://www.cellml.org/cellml/2.0#\" xmlns:cellml=\"http://www.cellml.org/cellml/2.0#\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" name=\"sin\" id=\"sin\">\n"
+        "  <component name=\"sin\" id=\"sin\">\n"
+        "    <variable name=\"x\" units=\"dimensionless\" interface=\"public_and_private\"/>\n"
+        "    <variable id=\"sin\" units=\"dimensionless\" name=\"sin\" interface=\"public_and_private\"/>\n"
+        "    <math xmlns=\"http://www.w3.org/1998/Math/MathML\">\n"
+        "  <apply id=\"actual_sin\"><eq/>\n"
+        "    <ci>sin</ci>\n"
+        "    <apply><sin/>\n"
+        "      <ci>x</ci>\n"
+        "    </apply>\n"
+        "  </apply>\n"
+        "</math>\n"
```

Comment:
This is not "nice". I would expect the indentation to be right. Will try to "fix" it.

Reply:
Done.